### ISSUE TITLE
`Object(x) === x` fails to narrow to `object`

### ISSUE THREAD

**Issue by ljharb** [microsoft/TypeScript#57000]
### 🔎 Search Terms

narrow object

### 🕗 Version & Regression Information

I only tested this on latest, but I think it's always worked this way.

### ⏯ Playground Link

https://tsplay.dev/WKbDpm

### 💻 Code

```ts
var err: unknown;

if (Object(err) === err) {
	err.toString; // `err` type is unknown, should be `object`
}
```

### 🙁 Actual behavior

Errors because it thinks `err` is `unknown`

### 🙂 Expected behavior

Does not error because it knows `err` is `object`

### Additional information about the issue

`Object(x) === x` is the most idiomatic and reliable way to test if something is an object, since it avoids hardcoding typeof checks. It'd be great if it was narrowed here.

**Comment by Skalman** [comment]
Some more context, since I was confused initially.

These are equivalent (AFAICT):
```ts
Object(x) === x

(typeof x === "object" && x !== null) || typeof x === "function"
```

It is *not* equivalent with this, since `Object.create(null)` and objects from other realms only match the expressions above:
```ts
x instanceof Object
```

**Comment by ljharb** [comment]
Yes, that's correct. The difference is that if JS ever adds a new object `typeof` value, `Object(x) === x` will always work, while the typeof-based code will not.

(`instanceof` is never reliable for builtins and should be avoided)

**Comment by ljharb** [comment]
@RyanCavanaugh What more feedback would you want to see before deciding this is a good case to improve narrowing?

**Comment by RyanCavanaugh** [comment]
@ljharb hard to quantify, but zero upvotes in 18 months is not a great indicator for something with sufficient demand to warrant action here. Some kind of proof that this pattern is idiomatic "in the wild" would be other data that would be interesting. It seems like most people would use `instanceof SomeMoreUsefulCtor` in most cases and it's hard to reason about which APIs need exactly `object` like this. Even
```ts
if (err) {
	err.toString
```
works fine today

**Comment by ljharb** [comment]
It's quite difficult to search for patterns like that unfortunately, since the variable name, and the order of the two operands, will vary.

There's a lot of builtin APIs - on the web and in JS - that require a non-primitive, and anyone who wants to check before risking an exception would need a way to do it. The much more common way is probably `x && (typeof x === 'object' || typeof x === 'function')` (and most people forget to check functions).

In other words, I agree there may not be a lot of user demand for this, but unless it's very difficult to implement in narrowing, I'd hope that the TS team would be interested in making axiomatic checks like this more useful, even if the priority was low.

**Comment by RyanCavanaugh** [comment]
Complexity is a budget that needs to be actively managed at all times; part of the management is not implementing even unobjectionable changes unless there is a clearly-identifiable upside.
